lib/ecto/migration/index.ex

```diff
@@ -0,0 +1,192 @@
+          end
+        end
+        # remove all old indexes from the system table
+        from(i in Ecto.Migration.SystemTable.Index, where: i.tablename == ^tablename) |> repo.delete_all
+        # insert new indexes to the system table
+        all_indexes = Ecto.Migration.Index.get_all_indexes(module)
+        for {fields, opts} <- all_indexes do
+          {index_name, concurently, unique, index_type} = Ecto.Migration.Index.get_index_helper(opts, [])
+          repo.insert(%Ecto.Migration.SystemTable.Index{tablename: tablename, index: Enum.join(fields, ","),
+                                                        index_name: index_name, concurrently: concurently,
+                                                        unique: unique, index_type: index_type})
+        end
+        deletion
+    end
+  end
+
+  def get_all_indexes(module) do
+    case :erlang.function_exported(module, :get_indexes, 0) do
+      true ->
+        module.get_indexes()
+      _ ->
+        []
+    end
+  end
+
+  def is_index_updated([], _) do
+    false
+  end
+
+  def is_index_updated(_, nil) do
+    true
+  end
+
+  # Compare actual index from model with indexes stored in the
+  # system index table. Returns true if something changed.
+  def is_index_updated(all_actual_indexes, all_system_indexes) do
+    case is_index_deleted(all_system_indexes, all_actual_indexes) do
+      true ->
+        true
+      false ->
+        case is_new_index(all_system_indexes, all_actual_indexes) do
+          false ->
+            is_index_changed(all_system_indexes, all_actual_indexes)
+          _ ->
+            true
+        end
+    end
+  end
+
+  def is_new_index([], [index | all_actual_indexes]) do
+    true
+  end
+
+  def is_new_index([], _) do
+    false
+  end
+
+  def is_new_index([index_from_db | system_indexes], all_actual_indexes) do
+    case List.keyfind(all_actual_indexes, index_string_to_atom_list(index_from_db), 0) do
+      nil ->
+        # new index added
+        true
+      index ->
+        is_new_index(system_indexes, all_actual_indexes -- [index])
+    end
+  end
+
+  defp is_index_deleted([], _) do
+    false
+  end
+
+  defp is_index_deleted([index_from_db | system_indexes], all_actual_indexes) do
+    case List.keyfind(all_actual_indexes, index_from_db |> index_string_to_atom_list, 0) do
+      nil ->
+        # we have no actual index in the database, so indexes changed
+        true
+      _ ->
+        is_index_deleted(system_indexes, all_actual_indexes)
+    end
+  end
+
+  defp is_index_changed([], _) do
+    false
+  end
+
+  defp is_index_changed([index_from_db | system_indexes], all_actual_indexes) do
+    opts_in_db = [index_name: index_from_db.index_name,
+                  concurrently: index_from_db.concurrently,
+                  index_type: index_from_db.index_type,
+                  unique: index_from_db.unique] |> :lists.sort
+    case compare_opts(opts_in_db, all_actual_indexes) do
+      {true, all_actual_indexes} ->
```

Comment:
compare_opts overwrite the all_actual_indexes (name would be better new_indexes?) Is, that what you really want here?